src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py

```diff
@@ -51,45 +81,56 @@ def execute(
-                    self._dag_manager.complete_nodes(self._collect_outputs(completed_futures))
-                    for each_future in completed_futures:
-                        del self._future_to_node[each_future]
-                    if timeout_task and timeout_task.done():
-                        raise LineExecutionTimeoutError(self._context._line_number, line_timeout_sec)
-                    self._execute_nodes(executor)
-            except Exception as e:
-                err_msg = "Flow execution has failed."
-                if isinstance(e, LineExecutionTimeoutError):
-                    err_msg = f"Line execution timeout after {line_timeout_sec} seconds."
-                    self._context.cancel_node_runs(err_msg)
-                node_names = ",".join(node.name for node in self._future_to_node.values())
-                flow_logger.error(f"{err_msg} Cancelling all running nodes: {node_names}.")
-                for unfinished_future in self._future_to_node.keys():
-                    # We can't cancel running tasks here, only pending tasks could be cancelled.
-                    unfinished_future.cancel()
-                # Even we raise exception here, still need to wait all running jobs finish to exit.
-                raise e
-            finally:
-                # Cancel timeout task no matter the execution is finished or failed.
-                event.set()
+        )
+        self._execute_nodes(executor)
+        timeout_task = None
+        event = threading.Event()
+        if line_timeout_sec is not None:
+            timeout_task = executor.submit(self.wait_within_timeout, event, line_timeout_sec)
+        try:
+            while not self._dag_manager.completed():
+                if not self._future_to_node:
+                    raise NoNodeExecutedError("No nodes are ready for execution, but the flow is not completed.")
+                tasks_to_wait = list(self._future_to_node.keys())
+                if timeout_task is not None:
+                    tasks_to_wait.append(timeout_task)
+                completed_futures_with_wait, _ = futures.wait(
+                    tasks_to_wait, timeout=10, return_when=futures.FIRST_COMPLETED
+                )
+                completed_futures = [f for f in completed_futures_with_wait if f in self._future_to_node]
+                self._dag_manager.complete_nodes(self._collect_outputs(completed_futures))
+                for each_future in completed_futures:
+                    del self._future_to_node[each_future]
+                if timeout_task and timeout_task.done():
+                    raise LineExecutionTimeoutError(self._context._line_number, line_timeout_sec)
+                self._execute_nodes(executor)
+            event.set()
+            executor.shutdown()
+        except Exception as e:
+            # When meet exception, mark event to exit timeout task.
+            event.set()
```

Comment:
Why not shutdown here?

Reply:
If call shudown here, will wait for running threads finish.
Maybe add shudown(wait=False) is anyway better than current code?